plugin/trino-sqlserver/src/main/java/io/trino/plugin/sqlserver/SqlServerClient.java

```diff
@@ -441,6 +481,56 @@ public void setNull(PreparedStatement statement, int index)
         };
     }
 
+    public static ColumnMapping longTimestampColumnMapping(TimestampType type)
+    {
+        if (type.getPrecision() <= MAX_SHORT_PRECISION) {
+            return timestampColumnMapping(type);
+        }
+        return ColumnMapping.objectMapping(
+                type,
+                longTimestampReadFunction(type),
+                longTimestampWriteFunction());
+    }
+
+    private static ObjectReadFunction longTimestampReadFunction(TimestampType type)
+    {
+        return ObjectReadFunction.of(
+                LongTimestamp.class,
+                (resultSet, columnIndex) -> {
+                    LocalDateTime localDateTime = resultSet.getObject(columnIndex, LocalDateTime.class);
+                    Instant instant = localDateTime.atZone(UTC).toInstant();
+                    long epochMicros = instant.getEpochSecond() * MICROSECONDS_PER_SECOND + floorDiv(instant.getNano(), NANOSECONDS_PER_MICROSECOND);
+                    int picosOfMicro = (instant.getNano() % NANOSECONDS_PER_MICROSECOND) * PICOSECONDS_PER_NANOSECOND;
+                    verifyTimestamp(epochMicros, picosOfMicro, type.getPrecision());
+                    return new LongTimestamp(epochMicros, picosOfMicro);
+                });
+    }
+
+    private static ObjectWriteFunction longTimestampWriteFunction()
+    {
+        return ObjectWriteFunction.of(
+                LongTimestamp.class,
+                (statement, index, value) -> {
+                    long epochSeconds = floorDiv(value.getEpochMicros(), MICROSECONDS_PER_SECOND);
+                    int microOfSecond = floorMod(value.getEpochMicros(), MICROSECONDS_PER_SECOND);
+                    int nanosOfSecond = (microOfSecond * NANOSECONDS_PER_MICROSECOND) +
+                            value.getPicosOfMicro() / PICOSECONDS_PER_NANOSECOND;
+
+                    Instant instant = Instant.ofEpochSecond(epochSeconds, nanosOfSecond);
+                    statement.setObject(index, LocalDateTime.ofInstant(instant, UTC));
+                });
+    }
+
+    protected static void verifyTimestamp(long epochMicros, int picosOfMicro, int precision)
+    {
+        if (precision > MAX_SHORT_PRECISION) {
+            verify(picosOfMicro % POWERS_OF_TEN[MAX_PRECISION - precision] == 0, "Invalid value of picosOfMicro for precision %s: %s", precision, picosOfMicro);
```

Comment:
`verify(picosOfMicro == round(picosOfMicro, 12 - precision),`
(using `io.trino.spi.type.Timestamps#round`) and you don't need `POWERS_OF_TEN`

Reply:
Done, thanks!